Serializer/Construction/DoctrineObjectConstructor.php

```diff
@@ -0,0 +1,121 @@
+use JMS\Serializer\Metadata\ClassMetadata;
+use JMS\Serializer\VisitorInterface;
+
+/**
+ * DoctrineObjectConstructor constructor is a copy of the original JMS constructor with additional logic to
+ * convert data to correctly find entity.
+ * Note: It not works with composite identifiers.
+ *
+ * @copyright 2018 OpticsPlanet, Inc
+ * @author    Vlad Yarus <vladislav.yarus@intexsys.lv>
+ */
+class DoctrineObjectConstructor implements ObjectConstructorInterface
+{
+    const ON_MISSING_NULL = 'null';
+    const ON_MISSING_EXCEPTION = 'exception';
+    const ON_MISSING_FALLBACK = 'fallback';
+    /**
+     * @var string
+     */
+    private $fallbackStrategy;
+
+    private $managerRegistry;
+    private $fallbackConstructor;
+
+    /**
+     * Constructor.
+     *
+     * @param ManagerRegistry $managerRegistry Manager registry
+     * @param ObjectConstructorInterface $fallbackConstructor Fallback object constructor
+     * @param string $fallbackStrategy
+     */
+    public function __construct(ManagerRegistry $managerRegistry, ObjectConstructorInterface $fallbackConstructor, $fallbackStrategy = self::ON_MISSING_NULL)
+    {
+        $this->managerRegistry = $managerRegistry;
+        $this->fallbackConstructor = $fallbackConstructor;
+        $this->fallbackStrategy = $fallbackStrategy;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function construct(VisitorInterface $visitor, ClassMetadata $metadata, $data, array $type, DeserializationContext $context)
+    {
+        $oldData = $data;
+        $isJsoApi = false;
+        if (array_key_exists('data', $data) && array_key_exists('id', $data['data'])) {
+            $isJsoApi = true;
+            $data = $oldData['data']['attributes'] ?? [];
+        }
+        // Locate possible ObjectManager
+        $objectManager = $this->managerRegistry->getManagerForClass($metadata->name);
+
+        if (!$objectManager) {
+            // No ObjectManager found, proceed with normal deserialization
+            return $this->fallbackConstructor->construct($visitor, $metadata, $data, $type, $context);
+        }
+
+        // Locate possible ClassMetadata
+        $classMetadataFactory = $objectManager->getMetadataFactory();
+
+        if ($classMetadataFactory->isTransient($metadata->name)) {
+            // No ClassMetadata found, proceed with normal deserialization
+            return $this->fallbackConstructor->construct($visitor, $metadata, $data, $type, $context);
+        }
+
+        // Managed entity, check for proxy load
+        if (!is_array($data)) {
+            // Single identifier, load proxy
+            return $objectManager->getReference($metadata->name, $data);
+        }
+
+        // Fallback to default constructor if missing identifier(s)
+        $classMetadata = $objectManager->getClassMetadata($metadata->name);
+        if ($isJsoApi) {
+            $data[$classMetadata->getIdentifierFieldNames()[0]] = $oldData['data']['id'];
```

Comment:
@vlady777 please add a comment that we are not supporting Composite Primary Keys. Probably we should throw an exception if there are more then 1 ID field names

Reply:
This comment already added in the class docblock